numpy/lib/npyio.py

```diff
@@ -761,6 +760,54 @@ def floatconv(x):
         return asstr
 
 
+# not to be confused with the flatten_dtype we import...
+def _flatten_dtype_internal(dt):
+    """Unpack a structured data-type, and produce re-packing info."""
+    if dt.names is None:
+        # If the dtype is flattened, return.
+        # If the dtype has a shape, the dtype occurs
+        # in the list more than once.
+        shape = dt.shape
+        if len(shape) == 0:
+            return ([dt.base], None)
+        else:
+            packing = [(shape[-1], list)]
+            if len(shape) > 1:
+                for dim in dt.shape[-2::-1]:
+                    packing = [(dim*packing[0][0], packing*dim)]
+            return ([dt.base] * int(np.prod(dt.shape)), packing)
+    else:
+        types = []
+        packing = []
+        for field in dt.names:
+            tp, bytes = dt.fields[field]
+            flat_dt, flat_packing = _flatten_dtype_internal(tp)
+            types.extend(flat_dt)
+            # Avoid extra nesting for subarrays
+            if tp.ndim > 0:
+                packing.extend(flat_packing)
+            else:
+                packing.append((len(flat_dt), flat_packing))
+        return (types, packing)
+
+
+def _pack_items(items, packing):
+    """Pack items into nested lists based on re-packing info."""
+    if packing is None:
+        return items[0]
+    elif packing is tuple:
+        return tuple(items)
+    elif packing is list:
+        return list(items)
+    else:
+        start = 0
+        ret = []
+        for length, subpacking in packing:
+            ret.append(_pack_items(items[start:start+length], subpacking))
+            start += length
+        return tuple(ret)
```

Comment:
If these are lifted from internal helpers inside `_loadtxt`, we should prefix their names with `_loadtxt`

I think it's probably worth a comment explaining _why_ these are lifted to the top level, and the refcounting / performance problems that putting them inside `loadtxt` caused.

Reply:
done